bigtable/client/cell.h

```diff
@@ -0,0 +1,99 @@
+
+#include <absl/base/thread_annotations.h>
+#include <absl/strings/string_view.h>
+#include <bigtable/client/version.h>
+#include <mutex>
+#include <vector>
+
+namespace bigtable {
+inline namespace BIGTABLE_CLIENT_NS {
+/**
+ * The in-memory representation of a Bigtable cell.
+ *
+ * Bigtable stores data in rows, indexes by row keys.  Each row may contain
+ * multiple column families, each column family might contain multiple columns,
+ * and each column has multiple cells indexed by timestamp.  Notice that the
+ * storage is sparse, column families, columns, and timestamps might contain
+ * zero cells.
+ *
+ * The Cell class owns all its data. Its string accessors return
+ * absl::string_view, which is a lightweight pointer to the string contained in
+ * Cell and does not actually hold any data. If those values are needed beyond
+ * the lifetime of the Cell object itself, they need to be copied.
+ */
+class Cell {
+ public:
+  /// Create a Cell and fill it with data.
+  Cell(std::string row_key, std::string family_name,
+       std::string column_qualifier, int64_t timestamp,
+       std::vector<std::string> chunks, std::vector<std::string> labels)
+      : row_key_(std::move(row_key)),
+        family_name_(std::move(family_name)),
+        column_qualifier_(std::move(column_qualifier)),
+        timestamp_(timestamp),
+        chunks_(std::move(chunks)),
+        labels_(std::move(labels)) {}
+
+  /// Return the row key this cell belongs to. The returned value is not valid
+  /// after this object dies.
+  absl::string_view row_key() const { return row_key_; }
+  /// Return the family this cell belongs to. The returned value is not valid
+  /// after this object dies.
+  absl::string_view family_name() const { return family_name_; }
+  /// Return the column this cell belongs to. The returned value is not valid
+  /// after this object dies.
+  absl::string_view column_qualifier() const { return column_qualifier_; }
+
+  /// Return the timestamp of this cell.
+  int64_t timestamp() const { return timestamp_; }
+
+  /// Return the contents of this cell. The returned value is not valid after
+  /// this object dies.
+  absl::string_view value() const {
+    std::lock_guard<std::mutex> l(mu_);
+    if (not chunks_.empty()) {
+      consolidate();
+    }
+    return copied_value_;
+  };
+
+  /// Return the labels applied to this cell by label transformer read filters.
+  const std::vector<std::string>& labels() const { return labels_; }
+
+ private:
+  /// consolidate concatenates all the chunks and caches the resulting value. It
+  /// is safe to call it twice, after the first call it becomes a no-op.
+  void consolidate() const;
+
+  std::string row_key_;
+  std::string family_name_;
+  std::string column_qualifier_;
+  int64_t timestamp_;
+
+  mutable std::string copied_value_ GUARDED_BY(mu_);
+  mutable std::vector<std::string> chunks_ GUARDED_BY(mu_);
+  std::vector<std::string> labels_;
+
+  mutable std::mutex mu_;
```

Comment:
So one interesting question: do we need an internal mutex?  Should this class be like a STL collection, which are safe to use via `const` functions in multiple threads, but need external locking otherwise?
If we expect the majority of the users to use `Cell` objects in a single cell, then no mutex is the way to go.

Reply:
How about having both of a completely const Cell with values merged at construction and a ChunkedCell with lazy evaluated value(), e.g.:
```
Row row = // boat...
for (Cell const& cell : row.cells()) {
  // standard usage, cell is pre-consolidated and completely const
}

for (ChunkedCell const& cell : row.chunked_cells()) {
  // advanced usage
  chunks = cell.chunks();  // for smart zero-copy parsers
  value = cell.value();  // consolidates on first access, not thread-safe
}
```
This way users of cells() do not need to worry about thread safety, and users of chunked_cells() will hopefully read the docs carefully.